storagenode/piecestore/endpoint.go

```diff
@@ -551,6 +553,50 @@ func (endpoint *Endpoint) SaveOrder(ctx context.Context, limit *pb.OrderLimit, o
 	}
 }
 
+// Retain keeps only piece ids specified in the request
+func (endpoint *Endpoint) Retain(ctx context.Context, retainReq *pb.RetainRequest) (*pb.RetainResponse, error) {
+	peer, err := identity.PeerIdentityFromContext(ctx)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	filter, err := bloomfilter.NewFromBytes(retainReq.GetFilter())
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	limit := 20
+	offset := 0
+	piecesRemaining := true
+
+	for piecesRemaining {
+		pieceIDs, err := endpoint.pieceinfo.GetPiecesID(ctx, peer.ID, retainReq.GetCreationDate(), limit, offset)
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+		for _, pieceID := range pieceIDs {
+			if !filter.Contains(pieceID) {
+				err = errs.Combine(err, endpoint.store.Delete(ctx, peer.ID, pieceID))
+				err = errs.Combine(err, endpoint.pieceinfo.Delete(ctx, peer.ID, pieceID))
```

Comment:
This combining of errors can be dangerous. As there might be millions of piece IDs to be inspected, if thousands of them must be deleted and they fail with an error, then the combined error may become huge and crash the storage node with out of memory. Perhaps, we should just log the errors as they come.

Reply:
I have these two lines logging instead of combining now, but not sure if the other calls in the function (`GetPieceIDs`, `bloomfilter.NewFromBytes`, and `identity.PeerIdentityFromContext`) should have returned errors or logged errors too.